engine/runtime/src/main/java/org/enso/interpreter/runtime/data/atom/Atom.java

```diff
@@ -137,96 +155,179 @@ boolean hasMembers() {
-    if (members != null && members.contains(member)) {
-      return !isMethodProjectPrivate(type, member);
+  /** Get field getters for this atom's constructor. */
+  private Set<Function> getFieldGetters() {
+    var allMethods = constructor.getDefinitionScope().getMethodsForType(constructor.getType());
+    if (allMethods != null) {
+      return allMethods.stream()
+          .filter(method -> isFieldGetter(method) && isGetterForOwnField(method))
+          .collect(Collectors.toUnmodifiableSet());
     }
-    members = type.getDefinitionScope().getMethodNamesForType(type);
-    if (members != null && members.contains(member)) {
-      return !isMethodProjectPrivate(type, member);
+    return Set.of();
+  }
+
+  /**
+   * Returns true if the given {@code function} is a getter for a field inside this atom
+   * constructor.
+   *
+   * @param function the function to check.
+   * @return true if the function is a getter for a field inside this atom constructor.
+   */
+  private boolean isGetterForOwnField(Function function) {
+    if (function.getCallTarget() != null
+        && function.getCallTarget().getRootNode() instanceof GetFieldBaseNode getFieldNode) {
+      var fieldName = getFieldNode.getName();
+      var thisConsFieldNames =
+          Arrays.stream(constructor.getFields()).map(ArgumentDefinition::getName).toList();
+      return thisConsFieldNames.contains(fieldName);
     }
     return false;
   }
 
-  /** A member is readable if it is an atom constructor and if the atom constructor is public. */
+  private boolean isFieldGetter(Function function) {
+    return function.getCallTarget() != null
+        && function.getCallTarget().getRootNode() instanceof GetFieldBaseNode;
+  }
+
+  /** A member is invocable if it is readable, i.e., if it is a field or a method. */
+  @ExportMessage
+  @CompilerDirectives.TruffleBoundary
+  final boolean isMemberInvocable(String member) {
+    return isMemberReadable(member);
+  }
+
+  /** Readable members are fields of non-project-private constructors and public methods. */
   @ExportMessage
   @ExplodeLoop
   final boolean isMemberReadable(String member) {
+    for (int i = 0; i < constructor.getArity(); i++) {
+      if (member.equals(constructor.getFields()[i].getName())) {
+        return true;
+      }
+    }
+    var method = findMethod(member);
+    return method != null;
+  }
+
+  /**
+   * All methods are internal, including public methods. Fields of project-private constructor are
+   * internal as well.
+   */
+  @ExportMessage
+  final boolean isMemberInternal(String member) {
     if (hasProjectPrivateConstructor()) {
-      return false;
+      return true;
     }
     for (int i = 0; i < constructor.getArity(); i++) {
       if (member.equals(constructor.getFields()[i].getName())) {
-        return true;
+        // Fields of public constructor are not internal.
+        return false;
       }
     }
-    return false;
+    // All methods are internal.
```

Comment:
That appears to contradict `InteropLibrary.isMemberInternal` which says the default is `false`?

Reply:
This method returns `false` only for fields of a public constructor. All the other members, including all methods, are considered internal. That is, fields are "public members", methods are internal members. This is compatible with what we used to have before #9692.